### DISCUSSION TITLE
Why decode variable-type payloads as tuples of (type, value)?

### DISCUSSION THREAD

**Discussion by asmello** [python-sdbus/python-sdbus#15] [Q&A]
This library is amazing, it's the most ergonomic binding for DBus I've found for the Python language, by a long shot. Everything just makes sense and I'm constantly surprised by good design decisions. But there's one quirk that's been bothering me: why do we get Python tuples for variable-type values encoded in method returns and signals?

For example:
```python
bus = sdbus.sd_bus_open_system()
obj_manager = DbusObjectManagerInterfaceAsync.new_proxy("org.bluez", "/", bus=bus)
result = await obj_manager.get_managed_objects()
print(result["/org/bluez/hci0"]["org.bluez.Adapter1"]["Discoverable"])
```

Will print:
```python
('b', False)  # or ('b', True)
```

Clearly, python-sdbus is able to decode the response to the correct Python type, or the second element in the tuple wouldn't be typed as a `bool`. However, it still preserves the low-level DBus type code as the first element. I don't see the benefit of this approach - perhaps it's intended to help write code that handles different possible types in certain fields, like "if string, do this, if bool do that", but for that purpose one can simply use `type(value) is str` (or better yet `isinstance(value, str)`). Further, in this example the type is fixed, so there's never going to be a need for code like that. So to me it feels like this extra context is redundant and adds unnecessary friction. Am I missing something?

**Comment by igo95862** [comment]
Yes, variants are a bit of a pain.

Ultimately I chose solution based on pythons zen: 

"Explicit is better than implicit.": the D-Bus type is explicitly encoded.
" Simple is better than complex.": variants decoded same way as they are encoded.

I have a few ideas how to make it more pleasant to work with variants.

First, the variant tuple can be overloaded with helper methods or properties. For example, `variant.value`, `variant.type`. Also add some helper functions like `BoolVariant(True) -> ('b', True)`.

Second, the common structure in D-Bus is dictionary with string keys and variant values. (`a{sv}`) I might add a `@dataclass` like class to support such dictionaries. There is some work for NetworkManager with such style of dataclass.

> perhaps it's intended to help write code that handles different possible types in certain fields, like "if string, do this, if bool do that", but for that purpose one can simply use `type(value) is str` (or better yet `isinstance(value, str)`

There are 8 types of `int` and 3 types of `str` that are same type in Python but different types in D-Bus.

**Reply by asmello** [reply]
If the goal is to make the decoding reversible, I agree this is a good strategy. But how commonly is reversibility required? When I read values from DBus, I don't usually write them back. I write different values instead, which will usually come from other sources and start  off as normal Python objects. When I ask for an integer, I don't care if it was originally 8, 16 or 32 bits, all I care is that I can use it as a Python `int` (do arithmetic, etc.). When I write an integer, I want the library to figure out the right representation to use, which it should be able to do since it has access to the schema. Or maybe the goal is to avoid relying on introspection at runtime?

FWIW I think the tuples format is fine for internal representation, but it just feels out of place in a public API.

**Reply by asmello** [reply]
And BTW it doesn't need to be a dichotomy. I can probably put together a PR to make the tuple format opt-out, by setting a flag on the `.new_proxy()` call or globally. If that's agreeable.

**Reply by igo95862** [reply]
> When I write an integer, I want the library to figure out the right representation to use, which it should be able to do since it has access to the schema. Or maybe the goal is to avoid relying on introspection at runtime?

The inside types of a Variant are not actually a part of schema or introspection. You have to go look for some third party documentation to figure out what types does a specific daemon accepts for specific method.

> But how commonly is reversibility required? When I read values from DBus, I don't usually write them back.

This is how NetworkManager connection settings work. (read-update-write) Losing the type information would mean trying to write signed integer in to unsigned field.

We have recently started working on helper classes for those NetworkManager connection settings that have information about dbus types inside so the variant tuples could be avoided. (such classes also can be initialized as a simple Python classes and have type hints)

I would rather add functions that help with working on variants.

**Reply by asmello** [reply]
> The inside types of a Variant are not actually a part of schema or introspection.

Oh interesting, I could've sworn it was. What a missed opportunity. It makes a lot of sense then.

> I would rather add functions that help with working on variants.

Yes, I think implementing specialised classes and functions that help encode knowledge about the types in the code rather than the objects will be the way to go. :+1:

**Comment by igo95862** [comment]
There is also an unpacking trick:

```python3
_, is_discoverable = result["/org/bluez/hci0"]["org.bluez.Adapter1"]["Discoverable"]
print(is_discoverable)
```

**Reply by asmello** [reply]
Yeah, working around it is not difficult, but it's just a lot of friction when you just care about the value. For example, when I'm processing the payload from the `InterfaceAdded` signal in the bluez interface, I get the full device information, but it comes all packed in these variant structures. So I've had to write code to recurse into it and replace the variant-tuples with the values, just so I can relay the payload to higher level code for further processing. Feels like I'm reimplementing the parser.

**Reply by igo95862** [reply]
Can you post some code examples?

It would help me think about how to improve the situation with variants.

**Reply by igo95862** [reply]
I have an idea for `InterfacesAdded` helper that might make it much easier to work with. Make helper return pseudo objects meaning instead of yielding a dictionary it would yield objects made from interface class. (so the `object.attribute` syntax works, typing information preserved)

**Reply by asmello** [reply]
> Can you post some code examples?

Will find some, but not sure they are very informative.

> Make helper return pseudo objects meaning instead of yielding a dictionary it would yield objects made from interface class. (so the object.attribute syntax works, typing information preserved)

That sounds awesome! But didn't you say introspection didn't have the full variants schema? Is the idea to allow the user to provide a template class?

**Reply by asmello** [reply]
Here's a method I use:
```python
async def _listen_added(self) -> None:
    async for path, interfaces in self._obj_manager.interfaces_added:
        if BLUEZ_DEVICE_INTERFACE not in interfaces:
            continue
        device_interface = interfaces[BLUEZ_DEVICE_INTERFACE]
        if (
            "ServiceData" not in device_interface
            or HUE_SERVICE_UUID not in device_interface["ServiceData"][1]
        ):
            continue
        device = Device(
            id=path.replace("/org/bluez/", ""),
            alias=device_interface["Alias"][1],
            address=device_interface["Address"][1],
            connected=device_interface["Connected"][1],
            paired=device_interface["Paired"][1],
            rssi=device_interface["RSSI"][1],
        )
        await self._events_queue.put(DeviceEvent(DeviceEventType.ADDED, device))
```
But I don't think it illustrates my point well. Sure, having the naked values would have allowed me to skip all the `[1]`s and to simplify the second if to `if HUE_SERVICE_UUID not in device_interface.get("ServiceData", ())`, but that's minor.

What actually frustrated me was the fact that I had to visually inspect the payload to make sure I was extracting the right values. I had D-Feet open and was expecting to find a string, yet got a tuple instead, and was initially confused. I think some documentation would've made it better, but I'd still have questioned why the payload was like that when D-Feet could list the exact type.

**Reply by igo95862** [reply]
> That sounds awesome! But didn't you say introspection didn't have the full variants schema? Is the idea to allow the user to provide a template class?

`InterfacesAdded` in particular contains a dict with the interfaces and properties (if any) that have been added to the given object path. The idea is that we already have interfaces classes so might reuse them.

**Reply by igo95862** [reply]
This is how I think it should look like once I implement a helper method:

```python
async def _listen_added(self) -> None:
    async for path, cached_interface in self._obj_manager.interface_object_added(DeviceInterface):
        if HUE_SERVICE_UUID not in cached_interface.service_data:
             continue

        device = Device(
            id=path.replace("/org/bluez/", ""),
            alias=cached_interface.alias,
            address=cached_interface.address,
            connected=cached_interface.connected,
            paired=cached_interface.paried,
            rssi=cached_interface.rssi,
        )
        await self._events_queue.put(DeviceEvent(DeviceEventType.ADDED, device))
```

**Reply by asmello** [reply]
This looks great. There's probably something similar to be done about the `org.freedesktop.DBus.Properties.PropertiesChanged` signal.

**Reply by igo95862** [reply]
> This looks great. There's probably something similar to be done about the `org.freedesktop.DBus.Properties.PropertiesChanged` signal.

`PropertiesChanged` signal only returns the changed properties so building entire class won't be possible. However, it does need a helper method that returns values in python names.

**Comment by igo95862** [comment]
`InterfacesAdded` in first place is not a very convenient signal as it deals with raw dbus interfaces and properties names.

**Comment by igo95862** [comment]
@asmello the latest release `0.11.0` has a new helper function `sdbus.utils.parse_interfaces_added` which will parse signal data to the python classes and member names.

**Reply by asmello** [reply]
That's awesome, thanks for the update!
